src/to_wiki.js

```diff
@@ -118,23 +168,48 @@ const serializer = new WikiTextSerializer({
    },
 
    text(state, node) {
-      let marks = node.marks || []
+      let marks = node.marks.map(m => m.type.name)
+
+      let firstNeedle = function(haystack, needles) {
+         return needles.reduce((min, needle) =>
+          Math.min(min, haystack.indexOf(needle)), haystack.length)
+      }
 
       // If openMarks includes marks that do not exist on the current node,
       // close those marks before adding new marks and the inline text to
       // the output.
-      let toClose = state.currentlyOpenMarks.filter(mark => marks.indexOf(mark) < 0)
-      state.closeMarks(toClose)
-
+      const toClose = state.currentlyOpenMarks.
+       filter(openMark => marks.indexOf(openMark) < 0)
+      // We need to close all marks that were opened after the oldest mark we
+      // need to close, so that we don't get overlapping marks (i.e. `<i>italic
+      // <b>italic bold</i> bold</b>`, but in wiki text).
+      const earliestClose = firstNeedle(state.currentlyOpenMarks, toClose)
+
+      state.closeMarks(state.currentlyOpenMarks.slice(earliestClose))
```

Comment:
It closes right-to-left (see [L137](https://github.com/iFixit/prosemirror-wikitext/blob/12434cbb83ba1df1edbb9ba2eed231d083196a2e/src/to_wiki.js#L137)), so it will close in LIFO order.